presto-hive/src/main/java/com/facebook/presto/hive/HiveCoercionRecordCursor.java

```diff
@@ -367,4 +386,297 @@ public void coerce(RecordCursor delegate, int field)
+        private final List<Type> fromFieldTypes;
+        private final List<Type> toFieldTypes;
+        private final Coercer[] coercers;
+
+        public StructCoercer(TypeManager typeManager, HiveType fromHiveType, HiveType toHiveType)
+        {
+            requireNonNull(typeManager, "typeManage is null");
+            requireNonNull(fromHiveType, "fromHiveType is null").getType(typeManager);
+            this.toType = requireNonNull(toHiveType, "toHiveType is null").getType(typeManager);
+            this.fromFieldHiveTypes = getAllStructFieldHiveTypes(fromHiveType);
+            this.toFieldHiveTypes = getAllStructFieldHiveTypes(toHiveType);
+            this.fromFieldTypes = fromFieldHiveTypes.stream()
+                    .map(hiveType -> hiveType.getType(typeManager))
+                    .collect(Collectors.toList());
+            this.toFieldTypes = toFieldHiveTypes.stream()
+                    .map(hiveType -> hiveType.getType(typeManager))
+                    .collect(Collectors.toList());
+            this.coercers = new Coercer[toFieldHiveTypes.size()];
+            Arrays.fill(this.coercers, null);
+            for (int i = 0; i < Math.min(fromFieldHiveTypes.size(), toFieldHiveTypes.size()); i++) {
+                if (!fromFieldTypes.get(i).equals(toFieldTypes.get(i))) {
+                    coercers[i] = createCoercer(typeManager, fromFieldHiveTypes.get(i), toFieldHiveTypes.get(i));
+                }
+            }
+        }
+
+        @Override
+        public void coerce(RecordCursor delegate, int field)
+        {
+            if (delegate.isNull(field)) {
+                setIsNull(true);
+                return;
+            }
+            Block block = (Block) delegate.getObject(field);
+            BlockBuilder builder = toType.createBlockBuilder(new BlockBuilderStatus(), 1);
+            BlockBuilder rowBuilder = builder.beginBlockEntry();
+            for (int i = 0; i < toFieldTypes.size(); i++) {
+                if (i >= fromFieldTypes.size() || block.isNull(i)) {
+                    rowBuilder.appendNull();
+                }
+                else if (coercers[i] == null) {
+                    block.writePositionTo(i, rowBuilder);
+                    rowBuilder.closeEntry();
+                }
+                else {
+                    rewriteBlock(fromFieldHiveTypes.get(i), toFieldHiveTypes.get(i), fromFieldTypes.get(i), toFieldTypes.get(i), block, i, rowBuilder, coercers[i], field);
+                }
+            }
+            builder.closeEntry();
+            setObject(builder.build().getObject(0, Block.class));
+        }
+    }
+
+    private static void rewriteBlock(
+            HiveType fromFieldHiveType,
+            HiveType toFieldHiveType,
+            Type fromFieldType,
+            Type toFieldType,
+            Block block,
+            int position,
+            BlockBuilder builder,
+            Coercer coercer,
+            int field)
+    {
+        Object value = null;
```

Comment:
I eventually used a single `BridgingRecordCursor` per `HiveCoercionRecordCursor` and de-referring the `value` variable at the end of `rewriteBlock`. I don't know if we actually need to set the correct `Type` here since we wouldn't visit `getType` method. What's your opinion?